cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -82,21 +82,77 @@ class UniqueAction : public ActionBase {
 
   Status Reserve(const int64_t length) { return Status::OK(); }
 
-  void ObserveNull() {}
+  Status ObserveNull() { return Status::OK(); }
 
   template <class Index>
-  void ObserveFound(Index index) {}
+  Status ObserveFound(Index index) {
+    return Status::OK();
+  }
 
   template <class Index>
-  void ObserveNotFound(Index index) {}
+  Status ObserveNotFound(Index index) {
+    return Status::OK();
+  }
 
   Status Flush(Datum* out) { return Status::OK(); }
 
   std::shared_ptr<DataType> out_type() const { return type_; }
+
+  Status FlushFinal(Datum* out) { return Status::OK(); }
+};
+
+// ----------------------------------------------------------------------
+// Count values implementation (see HashKernel for description of methods)
+
+class ValueCountsAction : ActionBase {
+ public:
+  using ActionBase::ActionBase;
+
+  ValueCountsAction(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : ActionBase(type, pool), count_builder_(pool) {}
+
+  Status Reserve(const int64_t length) {
+    // builder size is independent of input array size.
+    return Status::OK();
+  }
+  Status Reset() {
+    count_builder_.Reset();
+    return Status::OK();
+  }
+
+  // Don't do anything on flush beceause we don't want to finalize the builder
+  // or incur the cost of memory copies.
+  Status Flush(Datum* out) { return Status::OK(); }
+
+  std::shared_ptr<DataType> out_type() const { return type_; }
+
+  // Return the counts corresponding the MemoTable keys.
+  Status FlushFinal(Datum* out) {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(count_builder_.FinishInternal(&result));
+    out->value = std::move(result);
+    return Status::OK();
+  }
+
+  Status ObserveNull() { return Status::OK(); }
+
+  template <class Index>
+  Status ObserveFound(Index slot) {
+    count_builder_[slot]++;
```

Comment:
I'm curious how efficient the code that this generates is

Reply:
In release things seem to get inlined (can't find ObserveFound anymore).  In debug:
```
0000000000513740 <_ZN5arrow7compute12_GLOBAL__N_117ValueCountsAction12ObserveFoundIiEEvT_>:
  513740:       55                      push   %rbp
  513741:       48 89 e5                mov    %rsp,%rbp
  513744:       48 83 ec 10             sub    $0x10,%rsp
  513748:       48 89 7d f8             mov    %rdi,-0x8(%rbp)
  51374c:       89 75 f4                mov    %esi,-0xc(%rbp)
  51374f:       48 8b 7d f8             mov    -0x8(%rbp),%rdi
  513753:       48 83 c7 18             add    $0x18,%rdi
  513757:       48 63 75 f4             movslq -0xc(%rbp),%rsi
  51375b:       e8 90 1f d5 ff          callq  2656f0 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@plt>
  513760:       48 8b 30                mov    (%rax),%rsi
  513763:       48 83 c6 01             add    $0x1,%rsi
  513767:       48 89 30                mov    %rsi,(%rax)
  51376a:       48 83 c4 10             add    $0x10,%rsp
  51376e:       5d                      pop    %rbp
  51376f:       c3                      retq   
```
```
00000000002656f0 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@plt>:
  2656f0:       ff 25 8a 24 9f 00       jmpq   *0x9f248a(%rip)        # c57b80 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@@Base+0x731e70>
  2656f6:       68 6d 13 00 00          pushq  $0x136d
  2656fb:       e9 10 c9 fe ff          jmpq   252010 <.plt>
```